File: 2021/09/22/main.go
package main

import "math"

func maxLength(arr []string) int {
	return explore(arr, 0, 0)
}

func explore(a []string, j int, u int) int {
	if j == len(a) {
		return 0
	}

	cu := 0
	s := a[j]
	v := true
	for i := 0; i < len(s); i++ {
		ex := cu & (1 << (s[i] - 'a'))
		if ex == 0 {
			cu |= 1 << (s[i] - 'a')
		} else {
			v = false
			break
		}
	}

	r := explore(a, j+1, u)
	if v && (u&cu) == 0 {
		r = int(math.Max(float64(r),
			float64(len(s)+explore(a, j+1, u|cu))))
	}

	return r
}

File: 2021/09/22/main.py
class Solution:
    def maxLength(self, arr: List[str]) -> int:
        dp = [set()]
        for st in arr:
            if len(set(st)) < len(st):
                continue
            st = set(st)
            for c in dp[:]:
                if st & c:
                    continue
                dp.append(st | c)

        return max(len(a) for a in dp)